astropy/utils/data.py

```diff
@@ -968,6 +979,18 @@ def download_file(remote_url, cache=False, show_progress=True, timeout=None):
         The timeout, in seconds.  Otherwise, use
         `astropy.utils.data.Conf.remote_timeout`.
 
+    sources : list of str, optional
+        If provided, a list of URLs to try to obtain the file from. The
+        result will be stored under the original URL. The original URL
```

Comment:
The reasoning here is that there is a "true URL" for a resource, that will be used to look it up. Mirrors/fallbacks are common, and at least in principle provide the same file, so it seemed sensible to index it under the "true" location. Looking at the bug reports, some people will know they can't access the "true" location (say because they know they won't have access to US government or military servers) and in fact trying to access those might conceivably cause problems. So how are they to ensure that astropy doesn't even try the "true" URL?

What I still feel is missing is a way for users to specify lists of mirrors in their config file. This would avoid the calling code needing to list the mirrors - if you're running the code somewhere with limited Net access you presumably want to remap all astropy's requests without having to dig into the code. The main obstacle is I didn't see a nice way to express nested lists in the astropy config file.

Reply:
Specifically, astropy's config file format supports comma-separated lists, so for each resource you could have a config item listing mirrors for it. But what I was hoping for was a way to specify dictionaries, so you could put in a general map: if a url starts with this, try obtaining it by substituting each thing in this list for the initial location. 

For sources, I suppose we could rename it to "fallbacks" and have the convention that including None in the list meant to omit the original location. That doesn't really cover people who want to try their mirrors first, or only after the original location. My intention is that users should not normally have to specify "sources" by hand, but that libraries (like PINT) that know where resources can be found can supply a list of mirrors. (PINT, for example, needs to use the NANOGrav mirror of the JPL ephemeris server because that server blocked requests from Travis.)

Reply:
I actually quite liked the `sources` approach, but I also see @pllim's point.

An alternative would be to make the `remote_url` a list, where the file always gets stored under the first item, and have an `avoid` or `exclude` argument that tells which urls not to try.

Reply:
I think remote_url/sources is a clear way to divide up the functionality: remote_url provides a unique name that the resource will be listed under in the cache, while sources gives the practical information of where to actually obtain the content.

I guess "exclude" would also be a list? How would one handle exclude entries that were not in the original list - they are probably typos? Or would there be a global "exclude" list that declared URL prefixes that were inaccessible? How would users provide locally-available alternatives?

What I'd really like is a way for the code itself to provide the "true" URL that the file will be indexed under, but for customization to be able to adjust the list of places that content is looked for. So far the best way I've come up to do this is for functions like get_package_data to have config options, something like data_server_urls, that list potentially multiple urls to pass in as sources. (Right now I just pass in the data server and the alternative URL for the data server, individually configurable options.)

Another option, unfortunately not config-file-only, would be to have some kind of data structure in astropy.utils.data that could be used to mark some locations as inaccessible, optionally providing alternatives, based on URL prefixes. This is hinted at in download_files_parallel with the sources dictionary. My idea is that this, if implemented here or in user code, would leave remote_url alone but use it to construct or manipulate the sources list. One could even retrofit it so that leaving sources at the default meant using the sources mapper, while specifying sources meant the user knew exactly where to look for content.

Reply:
Thanks for the detailed thoughts - it does sound that easiest is just to stick with what you have. It is programmatically also easy: the default for `sources` is just the url. @pllim: ok with you too?

Reply:
Sorry for the delay. I'll think about this some more. This is next on my list to review next, so hopefully you'll hear more from me soon. Thanks for your patience!

Reply:
The scheme itself sounds OK, but how does one *avoid* trying the first URL? (say the `.mil` one for `finals2000A`).  Though perhaps a second argument that is just `skip_first` would work.

Alternatively, we could flip the arguments as given now, there is a list and by default it picks the first for the name to store it under, but one can override that with something like `store_under` or so. That keeps the advantage of @pllim's scheme that for the normal case one just passes in a list, without having to think about cache naming.

Reply:
If the first URL is known to be broken or undesirable, you remove it from the `dataurls` listing in your config, no? `dataurls` is a config item, so downstream packages/users should have complete control over it. My point is, there would be no skipping, you remove what you do not want to use. Did I miss something? I feel like I did...

Reply:
I think what @aarchiba was aiming for is that in your cache the data would still be stored under a standard name (based on that first URL). It perhaps is a fair question whether that's actually needed (which is why I suggested the reversal of arguments - that way in can be passed in for those cases it is needed).

Reply:
Ah, the hashes... Didn't we decide to use SHA of the file content, which would be the same no matter where it is downloaded from? Or did I misunderstand that too? Sorry, I have been jumping between many things and I might have misremembered.

Reply:
So the file content is stored under a hash, but that just means we don't need to worry about coming up with safe and valid filenames. In practice what you have is a dictionary mapping URLs to file contents. The URL is how you find the content; it is so that everyone who wants the IERS A table can just ask for the "true" IERS A URL and get the same table without re-downloading. I don't know of a better way for people to be able to reliably find the content they're looking for unless we decide to make up unique names for each kind of content astropy might want to index. You can't name them by the hash because some URLs have contents that change from time to time. (If you *wanted* to name them by the hash, you could just create a bogus URL byhash://abcdef0123..../ for the things and index under that. Of course this only works for code that knows the hash for contents that will never change; this is already more or less doable at the higher level of astropy-specific data accessed through get_data and friends.) 

How you *get* the contents of a URL if it is not present in the cache is a separate question. The normal case is that it lives on the Web at the URL you want to index it by, and normally you can just use urllib to pull that down and stick it in the cache. But this isn't always possible. For example, the US Navy might be planning to take down the canonical URL for the next six months (say). Or you might be behind a (national?) firewall that doesn't ever let you access US military servers. Or you might not be able to download anything from the Internet at all. Or the canonical server might have blocked your machines because you cause too much load by requesting too many files so you had to set up your own mirror (this happened to NANOGrav). So there should be some other way to get things into the cache. Ideally the code that requests the IERS A table (say) would not need to know what that other way of stuffing the cache was.

The easiest "other" way to get things into the cache is to request them from a different URL. Or a list of fall-back different URLs. This is what sources is for, though annoyingly I haven't found a way to separate it completely from the call site. If you know ahead of time you can call download_file(true_url, cache=True, sources=[...]) yourself, then the code that just calls download_file(true_url, cache=True) will just pull it out of the cache.

If you have no internet access, you can use sneakernet to get the file onto your system then call import_file(true_url, ...). After that the code that calls download_file(true_url, cache=True) will just pull it out of the cache. You can even export the entire cache from a machine that has requested all the data and import it into your sneakernet machine. 

`dataurls`: that is only relevant for downloads from the astropy server and its mirrors. Doesn't help with the IERS. This lives above download_file and is implemented with sources= so that whether the file is obtained from the server or the mirror it is stored under the same key. I'm happy to generalize that mechanism but coming up with a design that handles other servers with other mirroring arrangements brings you to my idea of storing a dictionary of true_url->mirror list in the config file; this is hampered by the fact that the config file format cannot represent dictionaries or even (as far as I can tell) lists of lists.

`skip_first` doesn't do a great job of addressing the situation where you want the original server to be tried but only after all the mirrors (e.g. we really want it to be the NANOGrav server we hammer but if it's not available fall back to JPL in case we're on a machine they haven't blocked yet). Of course you can then create a list to pass in that has the true URL twice, once at the beginning but skipped because you set `skip_first` and then again later in the list.

Part of the confusion is that the `get_data` series of functions is specific to the astropy data server and its mirror, and I haven't messed with that because it does not seem feasible to put useful data on those. (They don't ever update as far as I can tell; there are ancient outstanding bugs for additions to the observatory lists.) For non-astropy data that might live anywhere, the download_file and related functions are necessary.

Reply:
@aarchiba - thanks for the detailed reply. I definitely agree we need this and we should not let perfection be the enemy of good. So, as a last option to make it a bit more like @pllim and I were hoping for: how about my suggestion of allowing `remote_url` being a list (i.e., be treated as `sources` now), and have a new argument `store_under` (or so) that allows you to override which `url` to store the result under (which defaults to the first entry in the list). Of course, that really just amounts to swapping the arguments that you have, but I think it is a little more logical, that for the generic case where you have a main server and some fall-backs, you just pass those in.

Reply:
I don't like that design as much, but maybe that's just because it's not my idea. I think it will work fine. I'll put that in this afternoon with the additional special case that if the remote_url is a string rather than a list the obvious thing happens. Should I signal an error or warning somehow if the store_as is the same value as the first element of remote_url? That is if it's redundant?

Reply:
No need to check for redundancy, I'd think. Indeed, it might well be handy in some circumstances (like `IERS_Auto`) to always have the `store_as` listed.

Reply:
Also: any suggestions on how to handle download_files_in_parallel? Right now I just allow sources to be a map {true_url: [sources]} but I'm not sure what a good interface would be for this new download_file interface.

Perhaps download_files_in_parallel gets a list, each element of which can be:
- a URL,
- a list of URLs, the true URL of which is the first one, or
- a pair (list of URLs, store_as)?

This does seem more confusing, and does not much resemble the interface for download_file. I could have a separate list for store_as, but then that has to be kept in sync and also requires all or nothing in terms of specifying store_as.

Reply:
Let me also point out the use case where a user knows they don't have network access:
```
download_file(remote_urls=[], cache=True, store_as="http://thingiactuallywant.mil/")
```
versus
```
download_file("http://thingiactuallywant.mil/", cache=True, sources=[])
```
in particular, "store_as" is needed even though you are not storing anything.

Reply:
Hmm, that last example is quite convincing. @pllim, shall we just go with that @aarchiba has? It seems clear that it covers the bases better than what we are coming up with...

Reply:
How do you recommend I handle get_readable_fileobj? Its first argument is currently a string or a file-like object, and if it's a URL, then that object is requested through download_file; currently you can provide sources or not through a sources argument. Clearly store_as doesn't make sense if this is being passed an open file object or a local file, and what do we do if get_readable_fileobj is passed a list of local files instead of URLs - just keep going down the list until one can be opened?

I have implemented the conversion of download_file and download_files_in_parallel and some of their tests but am trying to convert some of the get_data family of functions.